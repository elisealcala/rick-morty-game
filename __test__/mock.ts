export const mockCard = {
	__typename: 'Character' as 'Character',
	id: '5',
	name: 'Jerry Smith',
	status: 'Alive',
	species: 'Human',
	type: '',
	gender: 'Male',
	image: 'https://rickandmortyapi.com/api/character/avatar/5',
};
