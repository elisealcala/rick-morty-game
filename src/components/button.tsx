import { FC } from "react";

type ButtonProps = {
  variant?: 'primary' | 'secondary',
  text: string
}

export const Button: FC<ButtonProps> = ({ variant = 'primary', text }) => {
  const className = variant === 'primary' ? 'bg-[#A2F2F9]' : 'bg-[#D8E054]';

  return (
    <button className={`${className} p-4`}>
      {text}
    </button>
  )
}