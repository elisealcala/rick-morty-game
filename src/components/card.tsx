import { FC } from 'react';
import Image from 'next/image';
import { GetCharactersQuery } from '../../generated';

type CardProps = NonNullable<
	NonNullable<NonNullable<GetCharactersQuery['characters']>['results']>[number]
> & {
	blocked?: boolean;
	flipped?: boolean;
	onClick?: () => void;
};

export const Card: FC<CardProps> = ({
	image,
	name,
	status,
	species,
	blocked,
	flipped,
	onClick,
}) => {
	const className = flipped
		? 'bg-[#A2F2F9] cursor-pointer flex items-center'
		: 'bg-white';

	return (
		<div
			onClick={!blocked && flipped ? onClick : () => null}
			className={`w-full p-4 h-[260px] rounded-[8px] shadow-md ${className}`}
			data-testid="container"
		>
			{flipped ? (
				<div className="relative w-full h-[180px]">
					<Image
						fill
						src={'/ricky_morty_img.png'}
						alt="rick_morty"
						className="rounded"
						sizes="180px, 180px"
					/>
				</div>
			) : (
				<>
					<div className="relative w-full h-[180px]">
						<Image
							fill
							src={image ?? ''}
							alt={name ?? ''}
							className="rounded"
							sizes="180px, 180px"
							data-testid="image"
						/>
					</div>
					<h2 className="mt-2 font-bold leading-[1.1]" data-testid="name">
						{name}
					</h2>
					<p className="mt-1 text-[10px]" data-testid="status">
						{status} - {species}
					</p>
				</>
			)}
		</div>
	);
};
