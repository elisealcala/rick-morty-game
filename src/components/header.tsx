import Image from 'next/image'

export const Header = () => {
  return (
    <div className='relative sm:w-[500px] sm:h-[200px]'>
      <Image src="/logo-rick-morty.png" alt="logo header" fill />
    </div>
  )
}