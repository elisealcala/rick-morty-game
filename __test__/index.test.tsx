import { Card } from '@/components/card';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { mockCard } from './mock';
import { Header } from '@/components/header';
import { Button } from '@/components/button';

describe('Card Component', () => {
	test('render with data', async () => {
		render(<Card {...mockCard} />);

		expect(screen.getByTestId('name')).toHaveTextContent(mockCard.name);
		expect(screen.getByTestId('status')).toHaveTextContent(
			`${mockCard.status} - ${mockCard.species}`
		);
	});

	test('avoid clicking when blocked', async () => {
		const onClick = jest.fn();

		render(<Card {...mockCard} onClick={onClick} flipped blocked />);

		fireEvent.click(screen.getByTestId('container'));

		expect(onClick).toHaveBeenCalledTimes(0);
	});

	test('it should fire the event when card is flipped', async () => {
		const onClick = jest.fn();

		render(<Card {...mockCard} onClick={onClick} flipped />);

		fireEvent.click(screen.getByTestId('container'));

		expect(onClick).toHaveBeenCalledTimes(1);
	});
});

describe('Header Component', () => {
	test('render correctly', async () => {
		const { container } = render(<Header />);

		expect(container).toMatchSnapshot();
	});
});

describe('Button Component', () => {
	test('render correctly', async () => {
		const { container } = render(<Button text="Example" />);

		expect(container).toMatchSnapshot();
	});

	test('fire onClick event', async () => {
		const onClick = jest.fn();

		render(<Button text="Example" onClick={onClick} />);

		fireEvent.click(screen.getByTestId('container'));

		expect(onClick).toHaveBeenCalledTimes(1);
	});
});
