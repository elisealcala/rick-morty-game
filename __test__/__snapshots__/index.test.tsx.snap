// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component render correctly 1`] = `
<div>
  <button
    class="bg-[#A2F2F9] border-[#D8E054]  border-b-4 border-l-4 px-[80px] py-3 rounded-[8px] font-semibold capitalize text-[24px] leading-[29px] text-center"
    data-testid="container"
  >
    Example
  </button>
</div>
`;

exports[`Header Component render correctly 1`] = `
<div>
  <div
    class="flex flex-col items-center"
  >
    <div
      class="relative w-[300px] h-[100px] sm:w-[500px] sm:h-[200px]"
    >
      <img
        alt="logo header"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="500px, 200px"
        src="/_next/image?url=%2Flogo-rick-morty.png&w=3840&q=75"
        srcset="/_next/image?url=%2Flogo-rick-morty.png&w=16&q=75 16w, /_next/image?url=%2Flogo-rick-morty.png&w=32&q=75 32w, /_next/image?url=%2Flogo-rick-morty.png&w=48&q=75 48w, /_next/image?url=%2Flogo-rick-morty.png&w=64&q=75 64w, /_next/image?url=%2Flogo-rick-morty.png&w=96&q=75 96w, /_next/image?url=%2Flogo-rick-morty.png&w=128&q=75 128w, /_next/image?url=%2Flogo-rick-morty.png&w=256&q=75 256w, /_next/image?url=%2Flogo-rick-morty.png&w=384&q=75 384w, /_next/image?url=%2Flogo-rick-morty.png&w=640&q=75 640w, /_next/image?url=%2Flogo-rick-morty.png&w=750&q=75 750w, /_next/image?url=%2Flogo-rick-morty.png&w=828&q=75 828w, /_next/image?url=%2Flogo-rick-morty.png&w=1080&q=75 1080w, /_next/image?url=%2Flogo-rick-morty.png&w=1200&q=75 1200w, /_next/image?url=%2Flogo-rick-morty.png&w=1920&q=75 1920w, /_next/image?url=%2Flogo-rick-morty.png&w=2048&q=75 2048w, /_next/image?url=%2Flogo-rick-morty.png&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <button
      class="mt-4 bg-[#D8E054] rounded-full py-2 px-4 text-[20px] leading-[24px] font-bold"
    >
      Juego de memoria
    </button>
  </div>
</div>
`;
